cpp/src/arrow/compute/row/grouper_benchmark.cc

```diff
@@ -0,0 +1,127 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/key_value_metadata.h"
+#include "arrow/util/string.h"
+#include "benchmark/benchmark.h"
+
+#include "arrow/compute/row/grouper.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/random.h"
+#include "arrow/util/benchmark_util.h"
+
+namespace arrow {
+namespace compute {
+
+constexpr auto kSeed = 0x0ff1ce;
+constexpr int64_t kRound = 256;
+
+static ExecBatch MakeRandomExecBatch(const DataTypeVector& types, int64_t num_rows,
+                                     double null_probability,
+                                     int64_t alignment = kDefaultBufferAlignment,
+                                     MemoryPool* memory_pool = nullptr) {
+  random::RandomArrayGenerator rng(kSeed);
+  auto num_types = static_cast<int>(types.size());
+
+  // clang-format off
+  auto metadata = key_value_metadata(
+      {
+        "null_probability",
+        "true_probability",
+        "unique"
+      },
+      {
+          internal::ToChars(null_probability),
+          internal::ToChars(null_probability),                     // for boolean type
+          internal::ToChars(static_cast<int32_t>(num_rows * 0.5))  // for string type
```

Comment:
That's gonna produce a lot of different groups if there's a string key, right? Is it realistic?

Reply:
This benchmark is mainly to ensure that the construction of our underlying RowTable and the performance of the comparison operations in the lower-level compare_internal can be tracked (we have not systematically tested these underlying operations before). It mainly covers some basics. Type, including the impact of null ratio on performance (comparison operations will compare null values separately.)

Reply:
Sure, but do we not want the workload to be more realistic?

Reply:
You are right, this is indeed a bit unrealistic. Can we use 0.1/0.2 to reduce the proportion of unique?

Reply:
Both scenarios have use cases.  GroupBy/HashJoin are sometimes used for deduplication and in that use case the key is mostly unique (in fact, we hit bugs in our hash-join impl when users tried this case).

That being said, I agree the "few groups" case is much more common.

Reply:
In addition, it seems that we can also enable swiss_join with the `are_cols_in_encoding_order=true` option, which can improve the performance of comparison when hashkey-cols need to be resorted.
https://github.com/apache/arrow/blob/7288bd557354677a9f1238ca652e528c5d522887/cpp/src/arrow/acero/swiss_join.cc#L245-L255

Reply:
> Both scenarios have use cases. GroupBy/HashJoin are sometimes used for deduplication and in that use case the key is mostly unique (in fact, we hit bugs in our hash-join impl when users tried this case).

I see, thanks. Can we clearly mark both kinds of scenarios in the benchmarks? (either as comments in the source code or by using different names for the benchmarks)

Reply:
Added the comments on Unique's impact scenarios in the benchmark.

But I think there is no need to add an independent benchmark for this. If the data set is all unique and there are non-performance bugs, this matter should be handled by a UT?

This Benchmark should focus on the impact of combining different data types and column rearrangements that occur in RowTable on comparison performance?